File: lira-playground/frontend/src/types/ast.ts
/**
 * Lira AST Type Definitions
 * Mirrors crates/lirac/src/ast.rs
 */

/** Source location information */
export interface Span {
  line: number;
  column: number;
}

/** Generic type parameter */
export interface TypeParam {
  name: string;
  bounds: string[];
}

/** A complete Lira program */
export interface Program {
  statements: Statement[];
}

/** Top-level and block statements */
export interface Statement {
  kind: StatementKind;
  span: Span;
}

/**
 * StatementKind types.
 *
 * Note: Some Rust variants are tuple variants (e.g., `Expression(Expression)`).
 * With serde's `tag = "type", content = "value"` and our normalizeAst() function,
 * tuple variants containing objects have their fields spread directly:
 * - Expression(expr) → {type: "Expression", kind: ExpressionKind, span: Span}
 * - Return(Some(expr)) → {type: "Return", kind: ExpressionKind, span: Span}
 * - Return(None) → {type: "Return", value: null}
 */
export type StatementKind =
  | { type: 'VarDecl'; pattern: Pattern; mutable: boolean; typeAnn: TypeExpr | null; initializer: Expression | null }
  | { type: 'ConstDecl'; name: string; typeAnn: TypeExpr | null; initializer: Expression }
  | { type: 'FnDecl'; name: string; typeParams: TypeParam[]; params: Parameter[]; returnType: TypeExpr | null; body: Block; isPublic: boolean; isOverride: boolean }
  | { type: 'ClassDecl'; name: string; parent: string | null; interfaces: string[]; fields: Field[]; methods: Statement[] }
  | { type: 'StructDecl'; name: string; typeParams: TypeParam[]; fields: Field[]; methods: Statement[] }
  | { type: 'EnumDecl'; name: string; variants: EnumVariant[] }
  | { type: 'InterfaceDecl'; name: string; methods: InterfaceMethod[] }
  | { type: 'TypeAlias'; name: string; typeExpr: TypeExpr }
  // Tuple variant: inner Expression fields are spread (kind, span)
  | { type: 'Expression'; kind: ExpressionKind; span: Span }
  // Tuple variant with Option: Some case has spread fields, None case has value: null
  | { type: 'Return'; kind?: ExpressionKind; span?: Span; value?: null }
  | { type: 'If'; condition: Expression; thenBranch: Block; elseBranch: Block | null }
  | { type: 'While'; condition: Expression; body: Block }
  | { type: 'For'; variable: string; iterable: Expression; body: Block }
  | { type: 'Loop'; body: Block }
  // Tuple variant with Option: Some case has spread fields, None case has value: null
  | { type: 'Break'; kind?: ExpressionKind; span?: Span; value?: null }
  | { type: 'Continue' }
  | { type: 'Import'; path: string[]; items: string[] | null }
  | { type: 'Use'; path: string[]; alias: string | null; items: string[] | null }
  | { type: 'TraitDecl'; name: string; typeParams: TypeParam[]; methods: TraitMethod[]; isPublic: boolean }
  | { type: 'ImplDecl'; traitName: string | null; typeName: string; typeParams: TypeParam[]; methods: Statement[] }
  | { type: 'Block'; block: Block };

/** A block of statements */
export interface Block {
  statements: Statement[];
  span: Span;
}

/** Function parameter */
export interface Parameter {
  name: string;
  typeAnn: TypeExpr;
  default: Expression | null;
  span: Span;
}

/** Struct/class field */
export interface Field {
  name: string;
  typeAnn: TypeExpr;
  isPublic: boolean;
  isMutable: boolean;
  span: Span;
}

/** Enum variant */
export interface EnumVariant {
  name: string;
  fields: TypeExpr[];
  span: Span;
}

/** Interface method signature */
export interface InterfaceMethod {
  name: string;
  params: Parameter[];
  returnType: TypeExpr | null;
  span: Span;
}

/** Trait method signature */
export interface TraitMethod {
  name: string;
  params: Parameter[];
  returnType: TypeExpr | null;
  hasSelf: boolean;
  defaultImpl: Block | null;
  span: Span;
}

/** Type expression */
export interface TypeExpr {
  kind: TypeExprKind;
  span: Span;
}

export type TypeExprKind =
  | { type: 'Named'; name: string }
  | { type: 'Generic'; name: string; args: TypeExpr[] }
  | { type: 'Optional'; inner: TypeExpr }
  | { type: 'Function'; params: TypeExpr[]; returnType: TypeExpr }
  | { type: 'Tuple'; elements: TypeExpr[] }
  | { type: 'Array'; element: TypeExpr }
  | { type: 'Result'; okType: TypeExpr; errType: TypeExpr }
  | { type: 'Path'; segments: string[] };

/** Expressions */
export interface Expression {
  kind: ExpressionKind;
  span: Span;
}

export type ExpressionKind =
  | { type: 'IntLiteral'; value: number }
  | { type: 'FloatLiteral'; value: number }
  | { type: 'StringLiteral'; value: string }
  | { type: 'CharLiteral'; value: string }
  | { type: 'BoolLiteral'; value: boolean }
  | { type: 'Null' }
  | { type: 'Identifier'; name: string }
  | { type: 'Binary'; left: Expression; op: BinaryOp; right: Expression }
  | { type: 'Unary'; op: UnaryOp; operand: Expression }
  | { type: 'Call'; callee: Expression; typeArgs: TypeExpr[]; args: Argument[] }
  | { type: 'FieldAccess'; object: Expression; field: string }
  | { type: 'OptionalAccess'; object: Expression; field: string }
  | { type: 'Index'; object: Expression; index: Expression }
  | { type: 'Array'; elements: Expression[] }
  | { type: 'Map'; entries: [Expression, Expression][] }
  | { type: 'Tuple'; elements: Expression[] }
  | { type: 'StructLiteral'; name: string | null; fields: [string, Expression][] }
  | { type: 'Lambda'; params: Parameter[]; body: Expression }
  | { type: 'IfExpr'; condition: Expression; thenExpr: Expression; elseExpr: Expression }
  | { type: 'Match'; subject: Expression; arms: MatchArm[] }
  | { type: 'Range'; start: Expression | null; end: Expression | null; inclusive: boolean }
  | { type: 'Cast'; expr: Expression; typeExpr: TypeExpr }
  | { type: 'TypeCheck'; expr: Expression; typeExpr: TypeExpr }
  | { type: 'Assign'; target: Expression; value: Expression }
  | { type: 'CompoundAssign'; target: Expression; op: BinaryOp; value: Expression }
  | { type: 'Block'; block: Block }
  | { type: 'Spawn'; expr: Expression }
  | { type: 'Select'; arms: SelectArm[] }
  | { type: 'EnumVariant'; enumName: string; variantName: string }
  | { type: 'Path'; segments: string[] }
  | { type: 'Try'; expr: Expression }
  | { type: 'MethodCall'; receiver: Expression; method: string; typeArgs: TypeExpr[]; args: Argument[] };

/** Binary operators */
export type BinaryOp =
  | 'Add' | 'Sub' | 'Mul' | 'Div' | 'Mod' | 'Pow'
  | 'Eq' | 'Ne' | 'Lt' | 'Le' | 'Gt' | 'Ge'
  | 'And' | 'Or'
  | 'BitAnd' | 'BitOr' | 'BitXor' | 'Shl' | 'Shr' | 'UShr'
  | 'NullCoalesce';

/** Unary operators */
export type UnaryOp = 'Neg' | 'Not' | 'BitNot' | 'PreInc' | 'PreDec' | 'PostInc' | 'PostDec';

/** Match arm */
export interface MatchArm {
  pattern: Pattern;
  guard: Expression | null;
  body: Expression;
  span: Span;
}

/** Function call argument */
export interface Argument {
  name: string | null;
  value: Expression;
  span: Span;
}

/** Select arm */
export interface SelectArm {
  kind: SelectArmKind;
  body: Expression;
  span: Span;
}

export type SelectArmKind =
  | { type: 'Recv'; variable: string | null; channel: Expression }
  | { type: 'Send'; value: Expression; channel: Expression }
  | { type: 'Default' };

/** Pattern for match expressions */
export interface Pattern {
  kind: PatternKind;
  span: Span;
}

export type PatternKind =
  | { type: 'Wildcard' }
  | { type: 'Literal'; expr: Expression }
  | { type: 'Variable'; name: string }
  | { type: 'Tuple'; elements: Pattern[] }
  | { type: 'Struct'; name: string; fields: [string, Pattern][]; rest: boolean }
  | { type: 'Constructor'; name: string; fields: Pattern[] }
  | { type: 'Range'; start: Pattern; end: Pattern; inclusive: boolean }
  | { type: 'Or'; patterns: Pattern[] }
  | { type: 'Binding'; name: string; pattern: Pattern };

/** Helper to get display name for statement kind */
export function getStatementKindName(kind: StatementKind): string {
  return kind.type;
}

/** Helper to get display name for expression kind */
export function getExpressionKindName(kind: ExpressionKind): string {
  return kind.type;
}

/** Helper to format binary operator */
export function formatBinaryOp(op: BinaryOp): string {
  const ops: Record<BinaryOp, string> = {
    Add: '+', Sub: '-', Mul: '*', Div: '/', Mod: '%', Pow: '**',
    Eq: '==', Ne: '!=', Lt: '<', Le: '<=', Gt: '>', Ge: '>=',
    And: '&&', Or: '||',
    BitAnd: '&', BitOr: '|', BitXor: '^', Shl: '<<', Shr: '>>', UShr: '>>>',
    NullCoalesce: '??',
  };
  return ops[op];
}

/** Helper to format unary operator */
export function formatUnaryOp(op: UnaryOp): string {
  const ops: Record<UnaryOp, string> = {
    Neg: '-', Not: '!', BitNot: '~',
    PreInc: '++', PreDec: '--', PostInc: '++', PostDec: '--',
  };
  return ops[op];
}

File: crates/lirac/src/ast.rs
//! Lira Abstract Syntax Tree
//!
//! Defines AST node types for the Lira parser.
//! See docs/lira/03-syntax-constructs.md for the full specification.

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

use crate::ids::NodeId;
use crate::lexer::Token;

/// Source location information
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Span {
    pub line: usize,
    pub column: usize,
}

impl From<&Token> for Span {
    fn from(token: &Token) -> Self {
        Span {
            line: token.line,
            column: token.column,
        }
    }
}

/// Generic type parameter (e.g., T in fn identity<T>)
#[derive(Debug, Clone, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct TypeParam {
    pub name: String,
    pub bounds: Vec<String>, // Trait bounds (e.g., T: Display)
}

/// A complete Lira program
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Program {
    pub statements: Vec<Statement>,
}

/// Top-level and block statements
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Statement {
    pub id: NodeId,
    pub kind: StatementKind,
    pub span: Span,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "type", content = "value"))]
pub enum StatementKind {
    /// Variable declaration: let x = expr or var x = expr
    /// Also supports destructuring: let (a, b) = tuple, let { x, y } = struct
    VarDecl {
        pattern: Pattern,
        mutable: bool,
        type_ann: Option<TypeExpr>,
        initializer: Option<Expression>,
    },
    /// Constant declaration
    ConstDecl {
        name: String,
        type_ann: Option<TypeExpr>,
        initializer: Expression,
    },
    /// Function declaration
    FnDecl {
        name: String,
        type_params: Vec<TypeParam>,
        params: Vec<Parameter>,
        return_type: Option<TypeExpr>,
        body: Block,
        is_public: bool,
        is_override: bool,
    },
    /// Class declaration
    ClassDecl {
        name: String,
        parent: Option<String>,
        interfaces: Vec<String>,
        fields: Vec<Field>,
        methods: Vec<Statement>,
    },
    /// Struct declaration
    StructDecl {
        name: String,
        type_params: Vec<TypeParam>,
        fields: Vec<Field>,
        methods: Vec<Statement>,
    },
    /// Enum declaration
    EnumDecl {
        name: String,
        variants: Vec<EnumVariant>,
    },
    /// Interface declaration
    InterfaceDecl {
        name: String,
        methods: Vec<InterfaceMethod>,
    },
    /// Type alias
    TypeAlias { name: String, type_expr: TypeExpr },
    /// Expression statement
    Expression(Expression),
    /// Return statement
    Return(Option<Expression>),
    /// If statement
    If {
        condition: Expression,
        then_branch: Block,
        else_branch: Option<Block>,
    },
    /// While loop
    While { condition: Expression, body: Block },
    /// For loop
    For {
        variable: String,
        iterable: Expression,
        body: Block,
    },
    /// Loop (infinite)
    Loop { body: Block },
    /// Break statement
    Break(Option<Expression>),
    /// Continue statement
    Continue,
    /// Import statement (legacy: import std.fs)
    Import {
        path: Vec<String>,
        items: Option<Vec<String>>,
    },
    /// Use statement (new: use std::fs)
    Use {
        path: Vec<String>,
        alias: Option<String>,
        items: Option<Vec<String>>,
    },
    /// Trait declaration
    TraitDecl {
        name: String,
        type_params: Vec<TypeParam>,
        methods: Vec<TraitMethod>,
        is_public: bool,
    },
    /// Impl block
    ImplDecl {
        trait_name: Option<String>,
        type_name: String,
        type_params: Vec<TypeParam>,
        methods: Vec<Statement>,
    },
    /// Block statement
    Block(Block),
}

/// A block of statements
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Block {
    pub statements: Vec<Statement>,
    pub span: Span,
}

/// Function parameter
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Parameter {
    pub id: NodeId,
    pub name: String,
    pub type_ann: TypeExpr,
    pub default: Option<Expression>,
    pub span: Span,
}

/// Struct/class field
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Field {
    pub name: String,
    pub type_ann: TypeExpr,
    pub is_public: bool,
    pub is_mutable: bool,
    pub span: Span,
}

/// Enum variant
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct EnumVariant {
    pub name: String,
    pub fields: Vec<TypeExpr>,
    pub span: Span,
}

/// Interface method signature
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct InterfaceMethod {
    pub name: String,
    pub params: Vec<Parameter>,
    pub return_type: Option<TypeExpr>,
    pub span: Span,
}

/// Trait method signature (with optional default implementation)
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct TraitMethod {
    pub name: String,
    pub params: Vec<Parameter>,
    pub return_type: Option<TypeExpr>,
    pub has_self: bool,
    pub default_impl: Option<Block>,
    pub span: Span,
}

/// Type expression
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct TypeExpr {
    pub kind: TypeExprKind,
    pub span: Span,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "type", content = "value"))]
pub enum TypeExprKind {
    /// Simple type name
    Named(String),
    /// Generic type: List<T>
    Generic { name: String, args: Vec<TypeExpr> },
    /// Optional type: T?
    Optional(Box<TypeExpr>),
    /// Function type: fn(A, B) -> C
    Function {
        params: Vec<TypeExpr>,
        return_type: Box<TypeExpr>,
    },
    /// Tuple type: (A, B, C)
    Tuple(Vec<TypeExpr>),
    /// Array type: [T]
    Array(Box<TypeExpr>),
    /// Result type: Result<T, E>
    Result {
        ok_type: Box<TypeExpr>,
        err_type: Box<TypeExpr>,
    },
    /// Path type: std::fs::File
    Path(Vec<String>),
    /// Inferred type — used for un-annotated function parameters.
    /// Resolves to `Type::Any` in the checker (the VM is dynamically typed).
    Infer,
}

/// Expressions
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Expression {
    pub id: NodeId,
    pub kind: ExpressionKind,
    pub span: Span,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "type", content = "value"))]
pub enum ExpressionKind {
    /// Integer literal
    IntLiteral(i64),
    /// Float literal
    FloatLiteral(f64),
    /// String literal
    StringLiteral(String),
    /// Character literal
    CharLiteral(char),
    /// Boolean literal
    BoolLiteral(bool),
    /// Null literal
    Null,
    /// Identifier
    Identifier(String),
    /// Binary operation
    Binary {
        left: Box<Expression>,
        op: BinaryOp,
        right: Box<Expression>,
    },
    /// Unary operation
    Unary {
        op: UnaryOp,
        operand: Box<Expression>,
    },
    /// Function call (supports named arguments and explicit type arguments)
    Call {
        callee: Box<Expression>,
        /// Explicit type arguments for generic functions (e.g., foo::<int, string>())
        type_args: Vec<TypeExpr>,
        args: Vec<Argument>,
    },
    /// Field access
    FieldAccess {
        object: Box<Expression>,
        field: String,
    },
    /// Optional chaining field access
    OptionalAccess {
        object: Box<Expression>,
        field: String,
    },
    /// Index access
    Index {
        object: Box<Expression>,
        index: Box<Expression>,
    },
    /// Array literal
    Array(Vec<Expression>),
    /// Map literal
    Map(Vec<(Expression, Expression)>),
    /// Tuple literal
    Tuple(Vec<Expression>),
    /// Struct/object literal
    StructLiteral {
        name: Option<String>,
        fields: Vec<(String, Expression)>,
    },
    /// Lambda expression
    Lambda {
        params: Vec<Parameter>,
        body: Box<Expression>,
    },
    /// If expression
    IfExpr {
        condition: Box<Expression>,
        then_expr: Box<Expression>,
        else_expr: Box<Expression>,
    },
    /// Match expression
    Match {
        subject: Box<Expression>,
        arms: Vec<MatchArm>,
    },
    /// Range expression
    Range {
        start: Option<Box<Expression>>,
        end: Option<Box<Expression>>,
        inclusive: bool,
    },
    /// Type cast
    Cast {
        expr: Box<Expression>,
        type_expr: TypeExpr,
    },
    /// Type check
    TypeCheck {
        expr: Box<Expression>,
        type_expr: TypeExpr,
    },
    /// Assignment
    Assign {
        target: Box<Expression>,
        value: Box<Expression>,
    },
    /// Compound assignment (+=, -=, etc.)
    CompoundAssign {
        target: Box<Expression>,
        op: BinaryOp,
        value: Box<Expression>,
    },
    /// Block expression
    Block(Block),
    /// Spawn expression
    Spawn(Box<Expression>),
    /// Select expression for multiple channel wait
    Select(Vec<SelectArm>),
    /// Enum variant access (Color::Red)
    EnumVariant {
        enum_name: String,
        variant_name: String,
    },
    /// Path expression (std::fs::read)
    Path { segments: Vec<String> },
    /// Try expression (expr?)
    Try(Box<Expression>),
    /// Method call with implicit self (receiver.method(args))
    MethodCall {
        receiver: Box<Expression>,
        method: String,
        /// Explicit type arguments for generic methods (e.g., obj.method::<int>())
        type_args: Vec<TypeExpr>,
        args: Vec<Argument>,
    },
}

/// Binary operators
#[derive(Debug, Clone, Copy, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub enum BinaryOp {
    // Arithmetic
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,
    // Comparison
    Eq,
    Ne,
    Lt,
    Le,
    Gt,
    Ge,
    // Logical
    And,
    Or,
    // Bitwise
    BitAnd,
    BitOr,
    BitXor,
    Shl,
    Shr,
    UShr,
    // Null coalescing
    NullCoalesce,
}

/// Unary operators
#[derive(Debug, Clone, Copy, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub enum UnaryOp {
    Neg,
    Not,
    BitNot,
    PreInc,  // ++x
    PreDec,  // --x
    PostInc, // x++
    PostDec, // x--
}

/// Match arm
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct MatchArm {
    pub pattern: Pattern,
    pub guard: Option<Expression>,
    pub body: Expression,
    pub span: Span,
}

/// Function call argument (supports named arguments)
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Argument {
    /// Optional parameter name for named arguments (e.g., `name:` in `foo(name: "bar")`)
    pub name: Option<String>,
    /// The argument value expression
    pub value: Expression,
    pub span: Span,
}

/// Select arm for channel select statement
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct SelectArm {
    pub kind: SelectArmKind,
    pub body: Expression,
    pub span: Span,
}

/// Type of select arm
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "type"))]
pub enum SelectArmKind {
    /// Receive from channel: variable = <-channel
    Recv {
        variable: Option<String>,
        channel: Expression,
    },
    /// Send to channel: value -> channel
    Send {
        value: Expression,
        channel: Expression,
    },
    /// Default case (non-blocking)
    Default,
}

/// Pattern for match expressions
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Pattern {
    pub id: NodeId,
    pub kind: PatternKind,
    pub span: Span,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "serde", serde(tag = "type", content = "value"))]
pub enum PatternKind {
    /// Wildcard pattern: _
    Wildcard,
    /// Literal pattern
    Literal(Expression),
    /// Variable binding
    Variable(String),
    /// Tuple pattern
    Tuple(Vec<Pattern>),
    /// Struct pattern
    Struct {
        name: String,
        fields: Vec<(String, Pattern)>,
        rest: bool,
    },
    /// Constructor/enum variant pattern: Some(x)
    Constructor { name: String, fields: Vec<Pattern> },
    /// Range pattern
    Range {
        start: Box<Pattern>,
        end: Box<Pattern>,
        inclusive: bool,
    },
    /// Or pattern
    Or(Vec<Pattern>),
    /// Binding pattern with @
    Binding { name: String, pattern: Box<Pattern> },
}
